python/pyarrow/dataset.py

```diff
@@ -232,104 +228,260 @@ def _ensure_format(obj):
-        a list of paths.
-    filesystem : FileSystem, default None
-        By default will be inferred from the path.
-    partitioning : Partitioning or PartitioningFactory or str or list of str
-        The partitioning scheme specified with the ``partitioning()``
-        function. A flavor string can be used as shortcut, and with a list of
-        field names a DirectionaryPartitioning will be inferred.
-    format : str, default None
-        Currently only "parquet" is supported.
+    paths : list of path-like
+        Note that URIs are not allowed.
+    filesystem : FileSystem or str, optional
+        If an URI is passed, then its path component will act as a prefix for
+        the file paths.
 
     Returns
     -------
-    FileSystemDatasetFactory
+    (FileSystem, list of str)
+        File system object and a list of normalized paths.
+
+    Raises
+    ------
+    TypeError
+        If the passed filesystem has wrong type.
+    FileNotFoundError
+        If the file system is local and a referenced path is not available.
+    ValueError
+        If the file system is local and a path references a directory or its
+        type cannot be determined.
     """
-    if not isinstance(path_or_paths, (list, tuple)):
-        path_or_paths = [path_or_paths]
-
-    partitioning = _ensure_partitioning(partitioning)
-    format = _ensure_format(format or "parquet")
-
-    # TODO pass through options
-    options = FileSystemFactoryOptions()
-    if isinstance(partitioning, PartitioningFactory):
-        options.partitioning_factory = partitioning
-    elif isinstance(partitioning, Partitioning):
-        options.partitioning = partitioning
-
-    factories = []
-    for path in path_or_paths:
-        fs, paths_or_selector = _ensure_fs_and_paths(path, filesystem)
-        factories.append(FileSystemDatasetFactory(fs, paths_or_selector,
-                                                  format, options))
-
-    if len(factories) == 0:
-        raise ValueError("Need at least one path")
-    elif len(factories) == 1:
-        return factories[0]
-    else:
-        return UnionDatasetFactory(factories)
-
-
-def _ensure_factory(src, **kwargs):
-    # Need to return DatasetFactory since `dataset` might need to finish the
-    # factory with a unified schema.
-    # TODO: return Dataset if a specific schema was passed?
-    if _is_path_like(src):
-        return factory(src, **kwargs)
-    elif isinstance(src, DatasetFactory):
-        if any(v is not None for v in kwargs.values()):
-            # when passing a SourceFactory, the arguments cannot be specified
-            raise ValueError(
-                "When passing a DatasetFactory, you cannot pass any "
-                "additional arguments"
-            )
-        return src
-    elif isinstance(src, Dataset):
+    if filesystem is None:
+        # fall back to local file system as the default
+        filesystem = LocalFileSystem()
```

Comment:
Another option to infer it from the first path? (re-using part of the code from `_ensure_single_source`)

Reply:
It doesn't support passing URIs as file paths. For example `['s3://file', 'file:///file']` wouldn't work. A dataset requires paths from the same filesystem.

Reply:
Yes, with two different filesystems it won't work (you will get an error then anyway). But if you pass multiple URIs from the same file system, it *could* work?

Reply:
May we defer this to a follow-up?

Reply:
Yes, that's certainly fine

Reply:
created https://issues.apache.org/jira/browse/ARROW-8400